src/jit/hwintrinsicxarch.cpp

```diff
@@ -1030,6 +1032,119 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
+
+                case TYP_LONG:
+                case TYP_ULONG:
+                {
+                    var_types shuffleBaseType = TYP_UNKNOWN;
+                    if (baseType == TYP_LONG)
+                    {
+                        intrinsic       = NI_SSE2_ConvertScalarToVector128Int64;
+                        shuffleBaseType = TYP_INT;
+                    }
+                    else
+                    {
+                        intrinsic       = NI_SSE2_ConvertScalarToVector128UInt64;
+                        shuffleBaseType = TYP_UINT;
+                    }
+#ifdef _TARGET_X86_
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_SetScalarVector128, TYP_DOUBLE, simdSize);
+#else
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, intrinsic, baseType, simdSize);
+#endif
+                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtNewIconNode(0b01000100), NI_SSE2_Shuffle,
+                                                       shuffleBaseType, simdSize);
+                    break;
+                }
+
+                case TYP_INT:
+                case TYP_UINT:
+                {
+                    src = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT,
+                                                   simdSize);
+                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtNewIconNode(0), NI_SSE2_Shuffle, baseType,
+                                                       simdSize);
+                    break;
+                }
+
+                case TYP_SHORT:
+                case TYP_USHORT:
+                {
+                    GenTree* srcClone = nullptr;
+
+                    // Initialize XMM register with initial value
+                    src = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT,
+                                                   simdSize);
+
+                    // src is used twice - clone it
+                    CORINFO_CLASS_HANDLE vectorHandle =
+                        (baseType == TYP_SHORT) ? Vector128ShortHandle : Vector128UShortHandle;
+
+                    srcClone = fgMakeMultiUse(&src, vectorHandle);
+
+                    GenTree* tmp =
+                        gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, srcClone, NI_SSE2_UnpackLow, TYP_SHORT, simdSize);
+                    retNode =
+                        gtNewSimdHWIntrinsicNode(TYP_SIMD16, tmp, gtNewIconNode(0), NI_SSE2_Shuffle, TYP_INT, simdSize);
+                    break;
+                }
+
+                case TYP_BYTE:
+                case TYP_UBYTE:
+                {
+                    GenTree* srcClone = nullptr;
+                    GenTree* tmpClone = nullptr;
+
+                    // Initialize XMM register with initial value
+                    src = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT,
```

Comment:
Same here, we need something that will zero/sign-extend `op1` before the `movd` instruction

Reply:
As above the codegen is identical to one used by MSVC and when I check disasm everything is taken care of:

for disasm see https://github.com/dotnet/coreclr/pull/16758#discussion_r172358008